[Enterprise Search Module][Behavioral Analytics] Event Intake API

afoucret opened the pull request:
This is a WIP draft PR for the vent intake branch.
The goal of this PR is to get  early feedback on the code.

## Description:

- [X] A new REST action is added to the code to receive event.

```
curl -H "Content-Type: application/json" -XPOST -u "elastic-admin:elastic-password" "localhost:9200/_application/analytics/bar/event/pageview" -d '

{
  "session": {
    "id": "dghdjdk"
  },
  "user": {
    "id": "bdjhdkdjdl"
  },
  "page": {
    "url": "http://elastic.co/blog"
  }
}'
```

- [x] The event is send into a log file where it can be consumed by filebeat 

## WIP
- [x] Unit Tests 70% written
 
## TODO LIST:
- [x] Licence check
- [ ] YAML rest tests
- [ ] API spec

## Will probably be done in a follow-up PR:
- [ ] Role for the intake API
- [ ] Default API key to receive events

tutelaris committed:
Introduce EventEmitterService

tutelaris committed:
Introduce SearchAnalyticsEvent

Commit:
A placeholder REST controller to push analytics events.

tutelaris committed:
Introduce InteractionAnalyticsEvent

tutelaris committed:
Introduce PageViewAnalyticsEvent

joemcelroy committed:
WIP

Commit:
Event emitter refactoring.

Commit:
Adding interaction events

Commit:
Updating data stream mapping.

Commit:
Refactoring to ease testing.

Commit:
Fix AnalyticsEvent serializing.

afoucret commented on x-pack/plugin/core/src/main/config/log4j2.properties:
```diff
@@ -104,6 +104,22 @@ appender.audit_rolling.policies.size.size = 1GB
 appender.audit_rolling.strategy.type = DefaultRolloverStrategy
 appender.audit_rolling.strategy.fileIndex = nomax
 
+appender.behavioral_analytics_rolling.type = RollingFile
```
ℹ️ A new logger is added to handle Analytics. The appender is just rendering the message string as received. We expect it to be called with the JSON event.

afoucret commented on x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/entsearch/analytics/behavioral_analytics-events-mappings.json:
ℹ️ Changed the mapping to be more consistent with what is sent through the tag

afoucret commented on x-pack/plugin/ent-search/build.gradle:
```diff
@@ -12,7 +12,8 @@ esplugin {
 
 dependencies {
-  compileOnly project(path: xpackModule('core'))
+  implementation project(path: ':libs:elasticsearch-x-content')
+    compileOnly project(path: xpackModule('core'))
```
⚠️ Remove ????

Commit:
Revert useless change in the module dependencies.

Commit:
Code refactoring.

Commit:
Adding tests.

tutelaris commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
@jimczi left a [comment](https://github.com/elastic/elasticsearch/pull/94404#discussion_r1132406852) in my PR that it's better to not use `toXContent` and replace it with `ESLogMessage` instead

jimczi replied:
I was missing the context in this comment. I think it's fine here since we need to parse the event for validation?

Commit:
Fixing forbidden API call.

Commit:
Added debug flag on the API.

Commit:
Fix failing test.

tutelaris commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEventParser.java:
```diff
@@ -0,0 +1,299 @@
+/*
```
I would probably have a class per event type, and create a factory on top of it that parses payload and creates an object accordingly.

Commit:
Refactored with one class per event type.

Commit:
Refactored with one class per event type.

Commit:
Revert debug code.

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java:
```diff
@@ -0,0 +1,152 @@
+
+            }
+            builder.endObject();
+
```
add session and user info

Commit:
Replacing interaction event type with search_click

joemcelroy committed:
update mapping

joemcelroy committed:
remove top level interaction

joemcelroy committed:
add document top level

joemcelroy committed:
add document top level

joemcelroy committed:
remove event.dataset

Commit:
A bunch of unit tests.

Commit:
Wrap resource init into a try

kderusso commented on x-pack/plugin/core/src/main/config/log4j2.properties:
```diff
@@ -104,6 +104,22 @@ appender.audit_rolling.policies.size.size = 1GB
+appender.behavioral_analytics_rolling.name = behavioral_analytics_rolling
+appender.behavioral_analytics_rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_analytics.json
+appender.behavioral_analytics_rolling.layout.type = PatternLayout
+appender.console.layout.pattern = %m%n
```
Is this line meant to be in the BA block?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java:
```diff
@@ -0,0 +1,248 @@
+
+        private final boolean accepted;
+
+        private final boolean debug;
```
Nitpick: Would a variable name such as `verbose` or `showEvent` be more clear than `debug`?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java:
```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
```
You could consider breaking this out into `DefaultResponse` and `DebugResponse`.

carlosdelest replied:
I like that idea. DebugResponse could be a subclass of DefaultResponse.

afoucret replied:
The class has been split in two different classes.
I did not used an `AcknowledgedResponse` because it is a response type that is meant to be used in the context of   master requests.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java:
```diff
@@ -0,0 +1,168 @@
+     * Analytics event types.
+     */
+    public enum Type {
+        PAGEVIEW("pageview"),
```
Would it make sense to rename this to `PAGE_VIEW`?

afoucret replied:
Changed in my new PR.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java:
```diff
@@ -0,0 +1,168 @@
+
+    private final AnalyticsEventUserData user;
+
+    protected AnalyticsEvent(String eventCollectionName, long eventTime, AnalyticsEventSessionData session, AnalyticsEventUserData user) {
```
Would it make sense to add a version field here, to preserve backward compatibility in future changes?

kderusso reviewed (commented):
Looking good so far!

Commit:
Add licence gate to the post event action.

Commit:
Refactoring

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java:
```diff
@@ -0,0 +1,248 @@
+            this.eventCollectionName = in.readString();
+            this.eventType = in.readString();
+            this.debug = in.readBoolean();
+            this.eventTime = in.readLong();
```
Are we including a timestamp for when the event has been received as well, which could be auto-generated by the server?

afoucret replied:
The constructor that is called by the rest handler is:

```
public Request(String eventCollectionName, String eventType, boolean debug, XContentType xContentType, BytesReference payload) {
   this(eventCollectionName, eventType, debug, System.currentTimeMillis(), xContentType, payload);
}
```

It generate the eventTime as soon as the request is received and use `System.currentTimeMillis()`.

The additional constructor are meant to be use for tests / deserialize request in the transport layer.

jimczi replied:
This makes me thinks that we should expect the eventTime to be set client side and be part of the sent event?

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java:
```diff
@@ -0,0 +1,248 @@
+            }
+
+            try {
+                if (eventType.toLowerCase(Locale.ROOT).equals(eventType) == false) throw new IllegalArgumentException();
```
Should we add a descriptive message for the exception?

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java:
```diff
@@ -0,0 +1,248 @@
+                AnalyticsEvent.Type.valueOf(eventType.toUpperCase(Locale.ROOT));
+            } catch (IllegalArgumentException e) {
+                validationException = addValidationError(
+                    LoggerMessageFormat.format("invalid event type: {}", eventType),
```
Including valid event types in message exception could be better for the user

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java:
```diff
@@ -0,0 +1,168 @@
+        this.eventCollectionName = Strings.requireNonBlank(eventCollectionName, "eventCollectionName");
+        this.eventTime = eventTime;
+        this.session = Objects.requireNonNull(session, AnalyticsEventSessionData.SESSION_FIELD.getPreferredName());
+        this.user = Objects.requireNonNull(user, AnalyticsEventUserData.USER_FIELD.getPreferredName());
```
I think it would be interesting to have, as part of the event information:
- An event ID, generated by the client, which can be used to deduplicate events if needed
- An auto-generated event received timestamp, which can be used to detect failures / bottlenecks in event processing, and not depend on the client timezone or timestamp generation. See [event-created](https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-created) ECS event field as an example.

Commit:
Lint.

carlosdelest reviewed (commented):
Some comments, looks really good!

Commit:
Fix error in the log4j config.

Commit:
Better unit tests.

joemcelroy commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java:
```diff
@@ -0,0 +1,276 @@
+
+    public static final PostAnalyticsEventAction INSTANCE = new PostAnalyticsEventAction();
+
+    public static final String NAME = "cluster:admin/xpack/application/analytics/post_event";
```
Should we update the privilege to not include admin?

joemcelroy commented:
Looks good so far! One thing I didn't see is we need CORS handling for the browser to send events to elasticsearch.

Commit:
Add tests for event classes.

Commit:
Finish unit tests.

Commit:
Fix error in AnalyticsEventFactoryTests.

Commit:
Rename EventEmitterService into AnalyticsEvent

Commit:
Add test for event type validation.

Commit:
Lint.

Commit:
Add a new cluster privilege to ingest events.

Commit:
Lint fix.

afoucret commented on x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc:
```diff
@@ -106,6 +106,7 @@ A successful call returns an object with "cluster" and "index" fields.
     "read_pipeline",
     "read_security",
     "read_slm",
+    "send_behavioral_analytics_events",
```
Typo here send_behavioral_analytics_event

afoucret replied:
Done in the new PR.

Commit:
Fix log4j2 config:
- behavioral_analytics appender should filter only events with the ANALYTICS_EVENT marker (uses a MarkerFilter)
- behavioral_analytics logger should not inherit root logger appenders (the analytics events should not be sent to main log)

jimczi commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/AnalyticsEventEmitterService.java:
```diff
@@ -0,0 +1,95 @@
+        final ActionListener<PostAnalyticsEventAction.Response> listener
+    ) {
+        try {
+            analyticsCollectionResolver.collection(clusterService.state(), request.eventCollectionName());
```
This can be removed?

jimczi replied:
Or maybe it is needed to check that the collection exists but the validation is missing?

tutelaris replied:
Afaik, here we indeed check if the collection is present. If not, it raises an exception, which we later catch and then respond with failure

jimczi reviewed (commented):
Thanks @afoucret ! This PR doesn't look like a draft ;).
The action is fine, I wonder though if a bulk mode is needed? Are we expecting the client to always send one event at a time?

We should also ask the core team to look at the log4j configuration. Looking at this big PR is maybe not the best for them so we should think of splitting the log4j part in a separate one?

Commit:
Small refactoring to avoid using markers to log event.
Also, unit tests are easier to read now.

Commit:
Implement most of the final events fields parsing.

Commit:
Simplify testing code.

afoucret commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:
```diff
@@ -268,6 +271,12 @@ public class ClusterPrivilegeResolver {
         MANAGE_BEHAVIORAL_ANALYTICS_PATTERN
     );
+
+    public static final NamedClusterPrivilege SEND_BEHAVIORAL_ANALYTICS_EVENT = new ActionClusterPrivilege(
```
Done in the new PR.

afoucret commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEventPageData.java:
```diff
@@ -0,0 +1,113 @@
+    );
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), PAGE_URL_FIELD);
```
Will be done in a follow-up PR.
